### ISSUE TITLE
Quantizing Bert

### ISSUE THREAD

**Issue by cjg91** [cjg91/trans-fat#3]
This issue tracks progress and understanding of the quantization problem. We will select a downstream task to fine-tune BERT and perform quantization-aware training. Then, we will take these quantized weights and implement the forward pass using pure Tensor functions to deobfuscate operations.

### Tasks:
- [x] Finetuning on IMDB in PyTorch -- [tutorial](https://huggingface.co/docs/transformers/training)
- [x] Dynamic quantization in PyTorch -- [tutorial](https://pytorch.org/tutorials/intermediate/dynamic_quantization_bert_tutorial.html)
- [x] Demonstrate dynamic quantization in numpy
- [ ] Get MLPerf quantize-aware model weights for BERT-Large
- [ ] Implement forward-pass for BERT-Large weights
- [ ] Train BERT with MLPerf QAT
- [ ] Try [Huggingface Optimum](https://github.com/huggingface/optimum) for static quantization

Questions:
- Q: Is the accumulator INT16 or INT32?
- A: INT32 seems [typical](https://arxiv.org/pdf/1712.05877.pdf) but intel uses [INT16 with saturation](https://oneapi-src.github.io/oneDNN/dev_guide_int8_computations.html)
- Q: How do you rescale to INT8 after matmul?
- A: Use fixed-point multiplication
- Q: How do you implement GeLU and softmax with INT8?
- Q: How does seq_len=128 affect accuracy?
- A: For the MRPC task everyone finetunes with max_seq_len=128

The fine-tuning tasks that I select:
MRPC
ML Perf uses SQuAD 1.1 as the downstream task. They also use BERT-Large, so we can't compare apples to apples.

The quantization technique I use:
TBD

Resources:
[PyTorch quantization intro](https://pytorch.org/blog/introduction-to-quantization-on-pytorch/)
[PyTorch quantization docs](https://pytorch.org/docs/stable/quantization.html)
[Quantization-supported operations](https://pytorch.org/docs/stable/quantization-support.html)
[BERT dynamic quantization Tutorial](https://pytorch.org/tutorials/intermediate/dynamic_quantization_bert_tutorial.html)
[Tensorflow quantized transformer paper](https://arxiv.org/pdf/1906.00532.pdf)
